Riot/Modules/MatrixKit/Models/Room/MXKRoomDataSourceManager.m

```diff
@@ -202,9 +225,19 @@ - (void)roomDataSourceForRoom:(NSString *)roomId create:(BOOL)create onComplete:
 
     // If not available yet, create the room data source
     MXKRoomDataSource *roomDataSource = roomDataSources[roomId];
-
+    
+    // check if the room's dataSource has events with late decryption issues and destroys it
+    BOOL roomDataSourceToBeDestroyed = [roomDataSourcesToDestroy containsObject:roomId];
+    
+    if (roomDataSource && roomDataSourceToBeDestroyed && create) {
+        [roomDataSource destroy];
+        roomDataSources[roomId] = nil;
+        roomDataSource = nil;
+    }
+    
```

Comment:
How often does this situation occur ? Because destroying every time seems like a rather costly action. 

Looking at `MXKRoomDataSource` code, I can see that there is a listener for decrypted event there. Can't we try to fix the dataSource content in `- (void)eventDidDecrypt:(NSNotification *)notif` rather ?